report/data_tasks/report/create_from_scratch/99_hubspot_write/02_contact_property_update/01_create_view.sql

```diff
@@ -2,18 +2,63 @@ DROP VIEW IF EXISTS hubspot.contact_property_update CASCADE;
 
 -- This view recreates the properties that we update on Hubspot contacts
 CREATE VIEW hubspot.contact_property_update AS (
+    WITH contact_users AS (
+        SELECT
+            contacts.id AS hs_object_id,
+            -- This is a little confusing, but the id here comes from H originally,
+            -- but is just processed through the LMS reporting system
+            MAX(users.id) AS h_user_id,
+            MIN(users.registered_date) AS lms_first_login
+        FROM hubspot.contacts
+        -- We are going to assume people never _stop_ being a teacher. This means
+        -- we will never remove the `Teacher` status from a contact. The upside is
+        -- we will update far fewer records, which should happen quicker.
+        JOIN lms.users ON
+            -- Emails in the contacts table are already stored as lower case
+            contacts.email = LOWER(users.email)
+            -- We definitely want to NOT join onto student information to keep that
+            -- out of Hubspot
+            AND users.is_teacher = True
+        GROUP BY contacts.id
+    )
+
     SELECT
-        contacts.id AS hs_object_id,
-        'Teacher' AS lms_role
-    FROM hubspot.contacts
-    -- We are going to assume people never _stop_ being a teacher. This means
-    -- we will never remove the `Teacher` status from a contact. The upside is
-    -- we will update far fewer records, which should happen quicker.
-    JOIN lms.users ON
-        -- Emails in the contacts table are already stored as lower case
-        contacts.email = LOWER(users.email)
-        -- We definitely want to NOT join onto student information to keep that
-        -- out of Hubspot
-        AND users.is_teacher = True
-    GROUP BY contacts.id
+        contact_users.hs_object_id,
+        contact_users.h_user_id,
+        'Teacher' AS lms_role,
+        contact_users.lms_first_login,
+        last_login.lms_last_login,
+        course_activity.lms_annotations_this_semester,
+        course_activity.lms_active_courses_this_semester
+    FROM contact_users
+
+    LEFT OUTER JOIN (
+        SELECT
+            user_id,
+            SUM(annotation_count) AS lms_annotations_this_semester,
+            COUNT(DISTINCT(group_roles.group_id)) AS lms_active_courses_this_semester
+        FROM lms.users
+        JOIN lms.group_roles ON
+            group_roles.user_id = users.id
+            AND group_roles.role = 'teacher'
+        JOIN lms.group_bubbled_activity ON
+            group_bubbled_activity.group_id = group_roles.group_id
+            AND group_bubbled_activity.created_week >= report.multi_truncate('semester', NOW()::date)
+        JOIN lms.groups ON
+            groups.id = group_roles.group_id
+        GROUP BY user_id
+    ) AS course_activity ON
+        course_activity.user_id = contact_users.h_user_id
+
+    LEFT OUTER JOIN (
```

Comment:
This is a bit weird. You'd think we'd do this with events, but we only have events since August 2022. The registered date is presumably the first time we saw them login, so it's probably the best we can do.

This leads to an odd imbalance with last login date. It results in users with a first login date, but without a last login (which is entirely event based).

We could try picking the `MAX(created)` from the LMS users and combine it with the data from the events table. I've not done this, as I've not pulled this data through, and it's an extra PR. I'd be interested to see if we can get away without it.

Reply:
You mentioned the other day that we have older launches in `lti_launches` but not from who. I forgot about this, we have 18M rows of launches without a user attached to them.